File: 스택/17298/17298.py
n = int(input())
s = []
arr = list(map(int, input().split()))
answer = [0] * n

for i in range(n - 1, -1, -1):
   while len(s) and s[-1] <= arr[i]:
      s.pop()
   answer[i] = s[-1] if len(s) else -1
   s.append(arr[i])
for i in answer:
   print(i, end=' ')

File: 스택/17298/17298.c
#include <stdio.h>
#include <string.h>
#define MAX 1000000

int main()
{
   int n, i, cnt = 0;
   int s[MAX], arr[MAX], answer[MAX];
   scanf("%d" , &n);
   for (i = 0; i < n; ++i)
      scanf("%d", arr + i);

   for (i = n - 1; i >= 0; --i)
   {
      while (cnt && s[cnt - 1] <= arr[i])
         --cnt;
      answer[i] = cnt ? s[cnt - 1] : -1;
      s[cnt++] = arr[i];
   }
   for (i = 0; i < n; ++i)
      printf("%d ", answer[i]);

   return 0;
}